### ISSUE TITLE
WIP: Training torch models in ML.NET

### ISSUE THREAD

**Issue by artidoro** [interesaaat/machinelearning#2]
First of all we need to distinguish between a TorchScript model and a Torch model that uses the C++ API. So I would suggest renaming the class `TorchModel` to `TorchScriptModel` and keep the name `TorchModel` for the torch models defined with the C++ API through C# bindings.

### Recap Torch Scoring Estimator
The `TorchScoringEstimator` currently requires the following arguments (where I replaced the current `TorchModel` with `TorchScriptModel`):
```csharp
            public string outputColumnName;
            public TorchScriptModel torchScriptModel;
            public long[] shapes;  // Input shapes.
            public string[] inputColumnNames;
```

### Torch Trainer
The torch trainer needs the following things in order to fully define the training process:
- the module defining parameters and the forward function `Torch.nn.Module` 
- the loss function
- the optimizer
- additional parameters such as: batch size, use gpu.

In ML.NET trainers are categorized based on the training task. We have many tasks: binary classification, multiclass classification, ranking, clustering, anomaly detection, regression.

One option is to add a new category that is task agnostic something like: DeepLearning.
Alternatively we would need to create a new extension for each learning task mentioned above. with little benefit as sometimes deep learning models might not fit well with the tasks existing in ML.NET. 

I would suggest for an initial version of this trainer to have it in a separate catalog which would make it easier for us. 

- **Possible Signature**

I would suggest the following signature for the `TorchTrainer`:
```csharp
public static Torch(this DeepLearningCatalog catalog, 
      Options options)
```
Where the `Options` class contains the following fields:
```csharp
public sealed class Options
{
            public string FeatureColumnName;
            public string LabelColumnName;
            public long[] FeatureColumnShape;
            public long[] LabelColumnShape;

            public TorchModel torchModel;
            public TorchLossFunction lossFunction;
            public TorchOptimizer optimizer;
            public bool useGpu;
            public int batchSize;
}
```

- **Possible Behavior**
```csharp
// Initialize the trainer estimator.
var options = new TorchTrainer.Options { ... };
var torchTrainer = mlContext.DeepLearning.Torch(options);

// Train.
var torchTransform = torchTrainer.Fit(data);
// Score.
var transformedData = torchTransform.Transform(data);
```

The transformed data should have the following new columns:
|`PredictedLabel`|`Score`|

Depending on the prediction and the associated loss function these could be probabilities or not. `Score` column would contain the loss. 

We still need to think about how to evaluate this type output in the ML.NET world, especially the difference between calibrated and non calibrated models.

**Comment by artidoro** [comment]
@wschin any thoughts?

**Comment by wschin** [comment]
We'd be to support multiple feature columns. Computation graph is a MIMO thing.

**Comment by interesaaat** [comment]
I believe MI is ok, but AFAIK we only have one output (the `forward` method in PyTorch modules return a single tensor).

No idea on how to do evaluation and calibration / not calibration (are these needed for probabilities?)

Regarding batching, do we really need it in the signature or we expect tensors already properly batched? (I personally think that batching should be an upstream problem of the featurization pipeline).
